src/MongoDbOperationStore.ts

```diff
@@ -0,0 +1,138 @@
+export class MongoDbOperationStore implements OperationStore {
+  private serverUrl: string;
+  private databaseName: string;
+  private collectionsName: string;
+  private client: MongoClient | undefined;
+  private db: Db | undefined;
+  private collection: Collection<any> | undefined;
+
+  constructor (config: Config) {
+    this.serverUrl = config[ConfigKey.OperationStoreUri];
+    this.databaseName = config[ConfigKey.OperationStoreDatabaseName];
+    this.collectionsName = config[ConfigKey.OperationStoreCollectionName];
+  }
+
+  /**
+   * Initialize the MongoDB operation store. The parameter resuming indicates
+   * whether the initialization is resuming from a previous stoppage with state
+   * stored in the backend mongodb database; if resuming is false, then a new
+   * database and collection is created.
+   */
+  public async initialize (resuming: boolean): Promise<void> {
+    this.client = await MongoClient.connect(this.serverUrl);
+    this.db = this.client.db(this.databaseName);
+
+    // If we are resuming, then the collection, indexes already exist.
+    // Otherwise, we need to create the collection and define an index on
+    // (did, transactionNumber, operationIndex), first dropping the collection if
+    // it exists.
+    if (resuming) {
+      this.collection = this.db.collection(this.collectionsName);
+    } else {
+      await this.db.collection(this.collectionsName).drop();
+      this.collection = await this.db.createCollection(this.collectionsName, { strict: false });
+      await this.collection.createIndex({ didUniqueSuffix: 1, transactionNumber: 1, operationIndex: 1 });
+    }
+  }
+
+  /**
+   * Implement OperationStore.put.
+   */
+  public async put (operation: Operation): Promise<void> {
+    const mongoOperation = this.getMongoOperation(operation);
+    await this.collection!.insertOne(mongoOperation);
+  }
+
+  /**
+   * Implement OperationStore.putBatch
+   */
+  public async putBatch (operations: Array<Operation>): Promise<void> {
+    let batch = this.collection!.initializeUnorderedBulkOp();
+
+    for (const operation of operations) {
+      const mongoOperation = this.getMongoOperation(operation);
+      batch.insert(mongoOperation);
+    }
+
+    await batch.execute();
+  }
+
+  /**
+   * Get an iterator that returns all operations with a given
+   * didUniqueSuffix ordered by (transactionNumber, operationIndex)
+   * ascending.
+   */
+  public async get (didUniqueSuffix: string): Promise<Iterable<Operation>> {
+    const mongoOperations = await this.collection!.find({ didUniqueSuffix }).sort({ transactionNumber: 1, operationIndex: 1 }).toArray();
+    return mongoOperations.map(this.getOperation);
+  }
+
+  /**
+   * Delete all operations with transaction number greater than the
+   * provided parameter.
+   */
+  public async delete (transactionNumber?: number): Promise<void> {
+    if (transactionNumber) {
+      await this.collection!.deleteMany({ transactionNumber: { $gt: transactionNumber } });
```

Comment:
Sanity: Are you sure there is no need to create a dedicated index for `transactionNumber` only in the `initialize` method?

Reply:
An additional index comes with a cost and delete has a specific access pattern (we never delete anything but recent operations). Let us discuss this today.